### PR TITLE
Dynamic Links

### PR DESCRIPTION
**Pull request by HardeepAsrani**
Closes https://github.com/Codeinwp/otter-internals/issues/26.

### Summary

This adds Dynamic Links to Otter's Dynamic Content module. It works the same way as Dynamic Content apart from having a shortcut.

Known issue:

- ~If you use both Text & Link on the same content, they might not work properly in the Backend.~
- ~If you use both Text & Link on the same content, they might not work properly in the Frontend.~

### Screenshots 

----

### Test instructions

Make sure all the conditions work the way you'd expect them and there are no PHP or JS errors. Some settings are free and some are Pro.

---- 

### Checklist before the final review

- [ ] Visual elements are not affected by independent changes.
- [ ] It is at least compatible with the [minimum WordPress version](https://wordpress.org/plugins/otter-blocks/).
- [ ] It loads additional script in frontend only if it is required.
- [ ] Does not impact the [Core Web Vitals](https://web.dev/vitals/).
- [ ] In case of deprecation, old blocks are safely migrated.
- [ ] It is usable in Widgets and FSE.

### CODE REVIEW

File: inc/plugins/class-dynamic-content.php [hunk 1]
```diff
@@ -45,6 +46,23 @@ public function apply_dynamic_content( $content ) {
...
+	 * @param string $content Post content.
+	 *
+	 * @return string
+	 */
+	public function apply_dynamic_link( $content ) { 
+		if ( false === strpos( $content, '<o-dynamic-link' ) ) {
+			return $content;
+		}
+
+		$re = '/<o-dynamic-link(?:\s+(?:data-type=["\'](?P<type>[^"\'<>]+)["\']|data-target=["\'](?P<target>[^"\'<>]+)["\']|data-meta-key=["\'](?P<metaKey>[^"\'<>]+)["\']|[a-zA-Z-]+=["\'][^"\'<>]+["\']))*\s*>(?<text>[^ $].*?)<\s*\/\s*o-dynamic-link>/';
```
Comment (Soare-Robert-Daniel) [inline inc/plugins/class-dynamic-content.php:61]: Do you have a tool for making those longs regexp, or just pure pain?

Comment (HardeepAsrani) [reply inc/plugins/class-dynamic-content.php:61]: It's not that big. It's a lot of repeated part. You can copy in Regex101 and see.

**Comment by HardeepAsrani** [comment]
After hours and 5-10 different methods of trying, I was finally able to fix how it works when both of the Dynamic options are used in a single chunk of text. It's not the best solution but it's the only one if 5-10 different options that worked.

Would love a review of the Dynamic Link module from you two @JohnPixle @mghenciu if you have time.

**Comment by irinelenache** [comment]
@HardeepAsrani Good job 🚀 Tested the feature and here are my findings: 

- [ ] The dynamic links don't work in the Advanced heading https://vertis.d.pr/C6XFwD
- [ ] If i added a **dynamic value** on a word, the option to add a **dynamic link** is still available, and after i select the link and hit Apply, nothing is happening and i'm getting some error in console https://vertis.d.pr/o0nICW (I think when dynamic value is added, the dynamic link should be disabled)
- [ ] Using **advanced custom fields**, if the field doesn't have the **URL type**, it will not be visible in the dynamic link dropdown https://vertis.d.pr/jXj844 (I think this is expected but i'm mentioning it just to be sure)
- [ ] The featured image link doesn't work when it is added in a custom layout (after header hook, after post content hook, before sidebar hook)

Besides these, everything seems to work fine for me 👍

**Comment by HardeepAsrani** [comment]
>  If i added a dynamic value on a word, the option to add a dynamic link is still available, and after i select the link and hit Apply, nothing is happening and i'm getting some error in console https://vertis.d.pr/o0nICW (I think when dynamic value is added, the dynamic link should be disabled)

We need to allow users to use both. For example they want to make a loop of Posts where the title leads to the post itself. I haven't been able to reproduce it still so can you post a video of showing how exactly it's happening?

>  The featured image link doesn't work when it is added in a custom layout (after header hook, after post content hook, before sidebar hook)

So the Featured Image doesn't work but the Author URL is working? In the same place, can you use the Post ID (Dynamic Value) and see if it gives you the ID of the post which is being viewed or ID of the Custom Layout?

**Comment by irinelenache** [comment]
@HardeepAsrani 

> If i added a dynamic value on a word, the option to add a dynamic link is still available, and after i select the link and hit Apply, nothing is happening and i'm getting some error in console https://vertis.d.pr/o0nICW (I think when dynamic value is added, the dynamic link should be disabled)

^ I checked this issue again and i managed to find why it was happening: I was adding a dynamic link without selecting a piece of text https://vertis.d.pr/PLAzlb

- [ ] Something similar is that dynamic link doesn't work alongside with the dynamic value, when the data type is the post id https://vertis.d.pr/oDMzNB

> The featured image link doesn't work when it is added in a custom layout (after header hook, after post content hook, before sidebar hook)

^ I gave more attention to this and all the dynamic values/links point to the custom layout (I was tricked by the fact that i'm the author of both the post and the custom layout 😅 )

**Comment by HardeepAsrani** [comment]
@irinelenache While I check the first issue, I suppose second issue should be fixed by https://github.com/Codeinwp/otter-blocks/pull/1170 - I'll test that and merge.

**Comment by HardeepAsrani** [comment]
> ^ I checked this issue again and i managed to find why it was happening: I was adding a dynamic link without selecting a piece of text https://vertis.d.pr/PLAzlb

This shouldn't happen anymore. I had left some code from Dynamic Text which was causing this.

**Comment by HardeepAsrani** [comment]
@irinelenache 

So there were some inconsistencies in how things were working when they were being used with Custom Layouts & Query Block that are now fixed. Here's how things should operate now (with Text, Link & Image). I've done testing to make sure it works the way I've described here, can you confirm the same?

Make sure you have the latest version, and the dynamic data is freshly created, the old one won't be consistent but shouldn't break either. Make sure you keep an eye on debug logs to make sure there are no errors.

### Dynamic Text & Link

- Post: Should get text/link based on the post being viewed.
- Reusable Block: Should get text/link based on the post being viewed.
- Query: It will show text/link based on the post of the query.
- Custom Layout in Neve: It will get the text/link based on the page being viewed.
- In Index/Archive Pages: Custom Layouts will fetch text/link from itself (not of index) while Query will just show the placeholder (no solution here for now because it’s loop within a loop). Things start to not working properly when it's a loop inside of a loop but I suppose at that far we can't do much.

### Dynamic Images

- Post: Should get image based on the post being edited or viewed, in both ends.
- Reusable Block: Should get image based on the post being edited or viewed, in both ends.
- Query: In the backend, it will show Placeholder while in the front, it will show image based on the post of the query.
- Custom Layout in Neve: While editing, it will get the image in the context of the Custom Layout being edited, while in the front it will get the image based on the page being viewed.
- In Index/Archive Pages: Custom Layouts will fetch image from itself (not of index) while Query will just show the placeholder (no solution here for now because it’s loop within a loop)

**Comment by irinelenache** [comment]
@HardeepAsrani I checked now and everything seems to work fine in the contexts you mentioned 🚀

**Pull request merged**